controllers/harbor/provision.go

```diff
@@ -0,0 +1,307 @@
+	namespacedName := harbor.getHarborCRNamespacedName()
+
+	return &v1alpha1.Harbor{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      namespacedName.Name,
+			Namespace: namespacedName.Namespace,
+			OwnerReferences: []metav1.OwnerReference{
+				*metav1.NewControllerRef(harbor.HarborCluster, goharborv1.HarborClusterGVK),
+			},
+		},
+		Spec: v1alpha1.HarborSpec{
+			HarborVersion: harbor.HarborCluster.Spec.Version,
+			PublicURL:     harbor.HarborCluster.Spec.PublicURL,
+			TLSSecretName: harbor.HarborCluster.Spec.TLSSecret,
+			Components: v1alpha1.HarborComponents{
+				Core:        harbor.newCoreComponentIfNecessary(),
+				Portal:      harbor.newPortalComponentIfNecessary(),
+				Registry:    harbor.newRegistryComponentIfNecessary(),
+				JobService:  harbor.newJobServiceComponentIfNecessary(),
+				ChartMuseum: harbor.newChartMuseumComponentIfNecessary(),
+				Clair:       harbor.newClairComponentIfNecessary(),
+				Notary:      harbor.newNotaryComponentIfNecessary(),
+			},
+			AdminPasswordSecret:  harbor.HarborCluster.Spec.AdminPasswordSecret,
+			Priority:             harbor.HarborCluster.Spec.Priority,
+			CertificateIssuerRef: harbor.HarborCluster.Spec.CertificateIssuerRef,
+		},
+	}
+}
+
+func (harbor *HarborReconciler) newCoreComponentIfNecessary() *v1alpha1.CoreComponent {
+	return &v1alpha1.CoreComponent{
+		HarborDeployment: v1alpha1.HarborDeployment{
+			Replicas:         IntToInt32Ptr(harbor.HarborCluster.Spec.Replicas),
+			Image:            harbor.ImageGetter.CoreImage(),
+			NodeSelector:     nil,
+			ImagePullSecrets: harbor.getImagePullSecrets(),
+		},
+	}
+}
+
+func (harbor *HarborReconciler) newPortalComponentIfNecessary() *v1alpha1.PortalComponent {
+	return &v1alpha1.PortalComponent{
+		HarborDeployment: v1alpha1.HarborDeployment{
+			Replicas:         IntToInt32Ptr(harbor.HarborCluster.Spec.Replicas),
+			Image:            harbor.ImageGetter.PortalImage(),
+			NodeSelector:     nil,
+			ImagePullSecrets: harbor.getImagePullSecrets(),
+		},
+	}
+}
+
+func (harbor *HarborReconciler) newRegistryComponentIfNecessary() *v1alpha1.RegistryComponent {
+	return &v1alpha1.RegistryComponent{
+		HarborDeployment: v1alpha1.HarborDeployment{
+			Replicas:         IntToInt32Ptr(harbor.HarborCluster.Spec.Replicas),
+			Image:            harbor.ImageGetter.RegistryImage(),
+			NodeSelector:     nil,
+			ImagePullSecrets: harbor.getImagePullSecrets(),
+		},
+		Controller: v1alpha1.RegistryControllerComponent{
+			Image: harbor.ImageGetter.RegistryControllerImage(),
+		},
+		StorageSecret: harbor.getRegistryStorageSecret(),
+		CacheSecret:   harbor.getRegistryCacheSecret(),
+	}
+}
+
+func (harbor *HarborReconciler) newJobServiceComponentIfNecessary() *v1alpha1.JobServiceComponent {
+	if harbor.HarborCluster.Spec.JobService != nil {
+		return &v1alpha1.JobServiceComponent{
+			HarborDeployment: v1alpha1.HarborDeployment{
+				Replicas:         IntToInt32Ptr(harbor.HarborCluster.Spec.Replicas),
```

Comment:
Do we support to specify separate replica value for some specified components such as job services?